In [1]:
pip install mlxtend

   ---------------------------------------- 0.0/1.4 MB ? eta -:--:--
   ---------------------------------------- 1.4/1.4 MB 30.2 MB/s  0:00:00
   ---------------------------------------- 0.0/12.4 MB ? eta -:--:--
   --------------------- ------------------ 6.8/12.4 MB 32.9 MB/s eta 0:00:01
   ---------------------------------------- 12.4/12.4 MB 33.9 MB/s  0:00:00

  Attempting uninstall: numpy

    Found existing installation: numpy 2.2.6

   ---------------------------------------- 0/2 [numpy]
    Uninstalling numpy-2.2.6:
   ---------------------------------------- 0/2 [numpy]
   ---------------------------------------- 0/2 [numpy]
   ---------------------------------------- 0/2 [numpy]
   ---------------------------------------- 0/2 [numpy]
   ---------------------------------------- 0/2 [numpy]
   ---------------------------------------- 0/2 [numpy]
   ---------------------------------------- 0/2 [numpy]
   ---------------------------------------- 0/2 [numpy]
   ------------------

  You can safely remove it manually.
  You can safely remove it manually.
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
opencv-python 4.12.0.88 requires numpy<2.3.0,>=2; python_version >= "3.9", but you have numpy 2.5.1 which is incompatible.
streamlit 1.50.0 requires pandas<3,>=1.4.0, but you have pandas 3.0.3 which is incompatible.
numba 0.61.0 requires numpy<2.2,>=1.24, but you have numpy 2.5.1 which is incompatible.
sklearn-compat 0.1.3 requires scikit-learn<1.7,>=1.2, but you have scikit-learn 1.9.0 which is incompatible.

[notice] A new release of pip is available: 26.0.1 -> 26.1.2
[notice] To update, run: C:\Users\teeva\anaconda3\python.exe -m pip install --upgrade pip


## 1. Import Necessary Libraries

In [2]:
import pandas as pd
import warnings

warnings.filterwarnings('ignore')

C:\Users\teeva\AppData\Roaming\Python\Python313\site-packages\pandas\core\computation\expressions.py:23: UserWarning: Pandas requires version '2.10.2' or newer of 'numexpr' (version '2.10.1' currently installed).
  from pandas.core.computation.check import NUMEXPR_INSTALLED


## 2. Data Collection

In [3]:
online_retail_data=pd.read_excel('Online Retail.xlsx')
online_retail_data

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
...,...,...,...,...,...,...,...,...
541904,581587,22613,PACK OF 20 SPACEBOY NAPKINS,12,2011-12-09 12:50:00,0.85,12680.0,France
541905,581587,22899,CHILDREN'S APRON DOLLY GIRL,6,2011-12-09 12:50:00,2.10,12680.0,France
541906,581587,23254,CHILDRENS CUTLERY DOLLY GIRL,4,2011-12-09 12:50:00,4.15,12680.0,France
541907,581587,23255,CHILDRENS CUTLERY CIRCUS PARADE,4,2011-12-09 12:50:00,4.15,12680.0,France


## 3. Data Understanding

In [4]:
online_retail_data.shape

(541909, 8)

In [5]:
online_retail_data.isna().sum()

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

In [6]:
online_retail_data.dtypes

InvoiceNo              object
StockCode              object
Description            object
Quantity                int64
InvoiceDate    datetime64[us]
UnitPrice             float64
CustomerID            float64
Country                   str
dtype: object

In [8]:
online_retail_data.describe(include='all')

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
count,541909.0,541909,540455,541909.000000,541909,541909.000000,406829.000000,541909
unique,25900.0,4070,4223,NaN,NaN,NaN,NaN,38
top,573585.0,85123A,WHITE HANGING HEART T-LIGHT HOLDER,NaN,NaN,NaN,NaN,United Kingdom
freq,1114.0,2313,2369,NaN,NaN,NaN,NaN,495478
mean,NaN,NaN,NaN,9.552250,2011-07-04 13:34:57.156386,4.611114,15287.690570,NaN
min,NaN,NaN,NaN,-80995.000000,2010-12-01 08:26:00,-11062.060000,12346.000000,NaN
25%,NaN,NaN,NaN,1.000000,2011-03-28 11:34:00,1.250000,13953.000000,NaN
50%,NaN,NaN,NaN,3.000000,2011-07-19 17:17:00,2.080000,15152.000000,NaN
75%,NaN,NaN,NaN,10.000000,2011-10-19 11:27:00,4.130000,16791.000000,NaN
max,NaN,NaN,NaN,80995.000000,2011-12-09 12:50:00,38970.000000,18287.000000,NaN


In [9]:
online_retail_data.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom


In [10]:
online_retail_data['Country'].unique()

<ArrowStringArray>
[      'United Kingdom',               'France',            'Australia',
          'Netherlands',              'Germany',               'Norway',
                 'EIRE',          'Switzerland',                'Spain',
               'Poland',             'Portugal',                'Italy',
              'Belgium',            'Lithuania',                'Japan',
              'Iceland',      'Channel Islands',              'Denmark',
               'Cyprus',               'Sweden',              'Austria',
               'Israel',              'Finland',              'Bahrain',
               'Greece',            'Hong Kong',            'Singapore',
              'Lebanon', 'United Arab Emirates',         'Saudi Arabia',
       'Czech Republic',               'Canada',          'Unspecified',
               'Brazil',                  'USA',   'European Community',
                'Malta',                  'RSA']
Length: 38, dtype: str

In [11]:
online_retail_data['Country'].value_counts() # no. of times occurring


Country
United Kingdom          495478
Germany                   9495
France                    8557
EIRE                      8196
Spain                     2533
Netherlands               2371
Belgium                   2069
Switzerland               2002
Portugal                  1519
Australia                 1259
Norway                    1086
Italy                      803
Channel Islands            758
Finland                    695
Cyprus                     622
Sweden                     462
Unspecified                446
Austria                    401
Denmark                    389
Japan                      358
Poland                     341
Israel                     297
USA                        291
Hong Kong                  288
Singapore                  229
Iceland                    182
Canada                     151
Greece                     146
Malta                      127
United Arab Emirates        68
European Community          61
RSA                         58


## Let's Build BEST ASSOCIATE For UK as it has more transactions

In [59]:
#saudi_data=online_retail_data[online_retail_data['Country'] == 'Saudi Arabia']
#saudi_data

In [13]:
uk_data=online_retail_data[online_retail_data['Country'] == 'United Kingdom']
uk_data

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
...,...,...,...,...,...,...,...,...
541889,581585,22466,FAIRY TALE COTTAGE NIGHT LIGHT,12,2011-12-09 12:31:00,1.95,15804.0,United Kingdom
541890,581586,22061,LARGE CAKE STAND HANGING STRAWBERY,8,2011-12-09 12:49:00,2.95,13113.0,United Kingdom
541891,581586,23275,SET OF 3 HANGING OWLS OLLIE BEAK,24,2011-12-09 12:49:00,1.25,13113.0,United Kingdom
541892,581586,21217,RED RETROSPOT ROUND CAKE TINS,24,2011-12-09 12:49:00,8.95,13113.0,United Kingdom


In [16]:
uk_data.isna().sum()

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     133600
Country             0
dtype: int64

In [18]:
1454/495478 

0.002934539979575279

In [17]:
del uk_data['CustomerID']

In [20]:
uk_data.dropna(inplace=True)

In [21]:
uk_data

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,United Kingdom
...,...,...,...,...,...,...,...
541889,581585,22466,FAIRY TALE COTTAGE NIGHT LIGHT,12,2011-12-09 12:31:00,1.95,United Kingdom
541890,581586,22061,LARGE CAKE STAND HANGING STRAWBERY,8,2011-12-09 12:49:00,2.95,United Kingdom
541891,581586,23275,SET OF 3 HANGING OWLS OLLIE BEAK,24,2011-12-09 12:49:00,1.25,United Kingdom
541892,581586,21217,RED RETROSPOT ROUND CAKE TINS,24,2011-12-09 12:49:00,8.95,United Kingdom


In [28]:
#saudi data
'''pd.pivot_table(data= saudi_data,
    values='Quantity',
    index='InvoiceNo',
    columns='Description',
    aggfunc='sum') '''

"pd.pivot_table(data= saudi_data,\n    values='Quantity',\n    index='InvoiceNo',\n    columns='Description',\n    aggfunc='sum') "

### Step 1:  turn the table into - transaction(row),products(column),value(quantity)

In [30]:
uk_pivotted_data=pd.pivot_table(data= uk_data,values='Quantity',index='InvoiceNo',columns='Description',aggfunc='sum').fillna(0)

In [31]:
uk_pivotted_data

Description,20713,4 PURPLE FLOCK DINNER CANDLES,50'S CHRISTMAS GIFT BAG LARGE,DOLLY GIRL BEAKER,I LOVE LONDON MINI BACKPACK,NINE DRAWER OFFICE TIDY,OVAL WALL MIRROR DIAMANTE,RED SPOT GIFT BAG LARGE,SET 2 TEA TOWELS I LOVE LONDON,SPACEBOY BABY GIFT SET,...,wrongly coded 20713,wrongly coded 23343,wrongly coded-23343,wrongly marked,wrongly marked 23343,wrongly marked carton 22804,wrongly marked. 23343 in box,wrongly sold (22719) barcode,wrongly sold as sets,wrongly sold sets
InvoiceNo,,,,,,,,,,,,,,,,,,,,,
536365,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
536366,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
536367,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
536368,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
536369,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
C581484,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
C581490,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
C581499,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0


### Step 2: change that table into 0,1 dataframe (one-hot df)

In [35]:
import pandas as pd
print(pd.__version__)


3.0.3


In [38]:
uk_pivotted_onehotdata=uk_pivotted_data.map(lambda x:0  if x==0 else 1)
# uk_pivotted_onehotdata = uk_pivotted_data.applymap(lambda x: 0 if x == 0 else 1)


# apply() - function only can apply to a column
# applymap() - function can apply to whole dataframe but it is in older version
# better use map()

In [39]:
uk_pivotted_onehotdata

Description,20713,4 PURPLE FLOCK DINNER CANDLES,50'S CHRISTMAS GIFT BAG LARGE,DOLLY GIRL BEAKER,I LOVE LONDON MINI BACKPACK,NINE DRAWER OFFICE TIDY,OVAL WALL MIRROR DIAMANTE,RED SPOT GIFT BAG LARGE,SET 2 TEA TOWELS I LOVE LONDON,SPACEBOY BABY GIFT SET,...,wrongly coded 20713,wrongly coded 23343,wrongly coded-23343,wrongly marked,wrongly marked 23343,wrongly marked carton 22804,wrongly marked. 23343 in box,wrongly sold (22719) barcode,wrongly sold as sets,wrongly sold sets
InvoiceNo,,,,,,,,,,,,,,,,,,,,,
536365,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
536366,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
536367,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
536368,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
536369,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
C581484,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
C581490,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
C581499,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0


## 4. Data Mining

In [15]:
from mlxtend.frequent_patterns import apriori,association_rules 
#apriori - to build best frequent set-items
#association rules - to build best association path (antecedent , consequent)

In [43]:
apriori_data = apriori(df=uk_pivotted_onehotdata,min_support=0.02,use_colnames=True) # df receive one-hot dataframe
apriori_data
# so, turn the table into - transaction(row),products(column),value(quantity)
# then, change that table into 0,1 dataframe

,support,itemsets
0,0.038884,frozenset({6 RIBBONS RUSTIC CHARM})
1,0.026497,frozenset({60 CAKE CASES VINTAGE CHRISTMAS})
2,0.034483,frozenset({60 TEATIME FAIRY CAKE CASES})
3,0.025363,frozenset({72 SWEETHEART FAIRY CAKE CASES})
4,0.040563,frozenset({ALARM CLOCK BAKELIKE GREEN})
...,...,...
242,0.021461,"frozenset({RED HANGING HEART T-LIGHT HOLDER, W..."
243,0.020962,"frozenset({RED RETROSPOT CHARLOTTE BAG, STRAWB..."
244,0.020962,"frozenset({REGENCY CAKESTAND 3 TIER, ROSES REG..."
245,0.024183,"frozenset({WOODEN FRAME ANTIQUE WHITE , WOODEN..."


In [47]:
best_associations =association_rules(df=apriori_data,metric='confidence',min_threshold=0.02,)
best_associations

,antecedents,consequents,antecedent support,consequent support,support,confidence,lift,representativity,leverage,conviction,zhangs_metric,jaccard,certainty,kulczynski
0,frozenset({ALARM CLOCK BAKELIKE RED }),frozenset({ALARM CLOCK BAKELIKE GREEN}),0.043512,0.040563,0.025817,0.593326,14.627420,1.0,0.024052,2.359232,0.974016,0.443146,0.576133,0.614896
1,frozenset({ALARM CLOCK BAKELIKE GREEN}),frozenset({ALARM CLOCK BAKELIKE RED }),0.040563,0.043512,0.025817,0.636465,14.627420,1.0,0.024052,2.631078,0.971022,0.443146,0.619928,0.614896
2,frozenset({CHARLOTTE BAG PINK POLKADOT}),frozenset({RED RETROSPOT CHARLOTTE BAG}),0.031352,0.041697,0.021915,0.698987,16.763518,1.0,0.020607,3.183593,0.970783,0.428571,0.685890,0.612279
3,frozenset({RED RETROSPOT CHARLOTTE BAG}),frozenset({CHARLOTTE BAG PINK POLKADOT}),0.041697,0.031352,0.021915,0.525571,16.763518,1.0,0.020607,2.041714,0.981262,0.428571,0.510216,0.612279
4,frozenset({RED RETROSPOT CHARLOTTE BAG}),frozenset({CHARLOTTE BAG SUKI DESIGN}),0.041697,0.036933,0.021688,0.520131,14.083142,1.0,0.020148,2.006936,0.969415,0.380876,0.501728,0.553677
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
85,"frozenset({GREEN REGENCY TEACUP AND SAUCER, RO...",frozenset({PINK REGENCY TEACUP AND SAUCER}),0.032350,0.033303,0.022550,0.697055,20.930634,1.0,0.021473,3.190995,0.984058,0.523158,0.686618,0.687083
86,"frozenset({PINK REGENCY TEACUP AND SAUCER, ROS...",frozenset({GREEN REGENCY TEACUP AND SAUCER}),0.025272,0.044011,0.022550,0.892280,20.274075,1.0,0.021438,8.874766,0.975325,0.482524,0.887321,0.702326
87,frozenset({GREEN REGENCY TEACUP AND SAUCER}),"frozenset({PINK REGENCY TEACUP AND SAUCER, ROS...",0.044011,0.025272,0.022550,0.512371,20.274075,1.0,0.021438,1.998913,0.994442,0.482524,0.499728,0.702326
88,frozenset({PINK REGENCY TEACUP AND SAUCER}),"frozenset({GREEN REGENCY TEACUP AND SAUCER, RO...",0.033303,0.032350,0.022550,0.677112,20.930634,1.0,0.021473,2.996856,0.985028,0.523158,0.666317,0.687083


In [52]:
import os
print(os.getcwd())


D:\Management\upm\SEM8\Paid_Data_Science\An Important_Syllabus\Machine Learning\CHAPTER20-ASSOCIATION RULES


In [58]:
#best_associations.to_csv("bestantecedentconsequent.csv") # restricted to save in working dir, so I paste the absolute path to save
best_associations.to_csv(r"C:\Users\teeva\OneDrive\Documents\best_antecedent_consequent.csv", index=False)

## Task

- apply this to another country
- learn about lift metric

### Lift
Lift = 1 (No Connection): Buying Item A has zero effect on Item B. It’s pure coincidence. (Example: Buying Milk and buying a Toothbrush. People need both, but buying milk doesn't make you suddenly want to brush your teeth).

Lift > 1 (Positive Connection): Buying Item A boosts the chances of buying Item B. They love each other! (Example: Burgers and Buns. If the Lift is 3, it means a customer is 3 times more likely to buy Buns if they already have Burgers in their cart).

Lift < 1 (Negative Connection): Buying Item A actually lowers the chances of buying Item B. They hate each other! (Example: If someone buys Coke, they are less likely to buy Pepsi at the same time).

A Simple Example (Let's do the Math!)Imagine we look at 100 shopping carts in your store:

50 carts have Burgers in them (50%).

40 carts have Buns in them (40%).

30 carts have Both Burgers and Buns (30%).

If Burgers and Buns had absolutely nothing to do with each other, we would expect them to end up in the same cart purely by chance. To find that "chance," we multiply their individual probabilities:$$\text{Chance} = 50\% \times 40\% = 20\%$$(So, by pure luck, we'd expect 20 out of 100 carts to have both).But in reality, 30% of carts have both!To find the Lift, we divide the real world by pure chance:$$\text{Lift} = \frac{\text{Real World Togetherness}}{\text{Pure Chance Togetherness}} = \frac{30\%}{20\%} = 1.5$$The Result:Because 1.5 is greater than 1, we have proven that Burgers and Buns are a great match. Specifically, having Burgers in the cart lifts the chances of buying Buns by 1.5 times (or 50% more than usual).